### PR TITLE
[Datasets] Allow `MultiHotEncoder` to encode arrays

### LINKED ISSUE #31353: [Datasets] `MultiHotEncoder` doesn't work with Arrow datasets
### What happened + What you expected to happen

Tried to transform a dataset with `MultiHotEncoder`, but got a `TypeError`:

```
Traceback (most recent call last):
  File "/Users/balaji/Documents/GitHub/ray/temp.py", line 7, in <module>
    encoder.fit_transform(dataset)
  File "/Users/balaji/Documents/GitHub/ray/python/ray/data/preprocessor.py", line 120, in fit_transform
    self.fit(dataset)
  File "/Users/balaji/Documents/GitHub/ray/python/ray/data/preprocessor.py", line 105, in fit
    return self._fit(dataset)
  File "/Users/balaji/Documents/GitHub/ray/python/ray/data/preprocessors/encoder.py", line 318, in _fit
    self.stats_ = _get_unique_value_indices(
  File "/Users/balaji/Documents/GitHub/ray/python/ray/data/preprocessors/encoder.py", line 550, in _get_unique_value_indices
    value_counts = dataset.map_batches(get_pd_value_counts, batch_format="pandas")
  File "/Users/balaji/Documents/GitHub/ray/python/ray/data/dataset.py", line 659, in map_batches
    return Dataset(plan, self._epoch, self._lazy)
  File "/Users/balaji/Documents/GitHub/ray/python/ray/data/dataset.py", line 223, in __init__
    self._plan.execute(allow_clear_input_blocks=False)
  File "/Users/balaji/Documents/GitHub/ray/python/ray/data/_internal/plan.py", line 321, in execute
    blocks, stage_info = stage(
  File "/Users/balaji/Documents/GitHub/ray/python/ray/data/_internal/plan.py", line 688, in __call__
    blocks = compute._apply(
  File "/Users/balaji/Documents/GitHub/ray/python/ray/data/_internal/compute.py", line 154, in _apply
    raise e from None
  File "/Users/balaji/Documents/GitHub/ray/python/ray/data/_internal/compute.py", line 138, in _apply
    results = map_bar.fetch_until_complete(refs)
  File "/Users/balaji/Documents/GitHub/ray/python/ray/data/_internal/progress_bar.py", line 75, in fetch_until_complete
    for ref, result in zip(done, ray.get(done)):
  File "/Users/balaji/Documents/GitHub/ray/python/ray/_private/client_mode_hook.py", line 105, in wrapper
    return func(*args, **kwargs)
  File "/Users/balaji/Documents/GitHub/ray/python/ray/_private/worker.py", line 2347, in get
    raise value.as_instanceof_cause()
ray.exceptions.RayTaskError(TypeError): ray::_map_block_split() (pid=82040, ip=127.0.0.1)
  File "/Users/balaji/Documents/GitHub/ray/python/ray/data/_internal/compute.py", line 459, in _map_block_split
    for new_block in block_fn(blocks, *fn_args, **fn_kwargs):
  File "/Users/balaji/Documents/GitHub/ray/python/ray/data/dataset.py", line 637, in transform
    yield from process_next_batch(batch)
  File "/Users/balaji/Documents/GitHub/ray/python/ray/data/dataset.py", line 601, in process_next_batch
    batch = batch_fn(batch, *fn_args, **fn_kwargs)
  File "/Users/balaji/Documents/GitHub/ray/python/ray/data/preprocessors/encoder.py", line 543, in get_pd_value_counts
    result[col] = get_pd_value_counts_per_column(df[col])
  File "/Users/balaji/Documents/GitHub/ray/python/ray/data/preprocessors/encoder.py", line 536, in get_pd_value_counts_per_column
    return Counter(col.value_counts(dropna=False).to_dict())
  File "/Users/balaji/Documents/GitHub/ray/.venv/lib/python3.10/site-packages/pandas/core/series.py", line 1895, in to_dict
    return into_c((k, maybe_box_native(v)) for k, v in self.items())
TypeError: unhashable type: 'numpy.ndarray'
```

### Versions / Dependencies

ray: f40ac957cbd41c4f7b3d999eaa9075119f958881
pyarrow: 10.0.1

### Reproduction script

```
import ray
from ray.data.preprocessors import MultiHotEncoder

dataset = ray.data.from_items([{"column": ["spam", "ham", "eggs"]}])
print(dataset)
encoder = MultiHotEncoder(columns=["column"])
encoder.fit_transform(dataset)
```

### Issue Severity

Medium: It is a significant difficulty but I can work around it.

### PR DESCRIPTION
**Pull request by bveeramani**
Signed-off-by: Balaji Veeramani <balaji@anyscale.com>

## Why are these changes needed?

**tl;dr**: `MultiHotEncoder` doesn't work with Arrow datasets. This PR updates the `MultiHotEncoder` implementation to fix the issue.

`MultiHotEncoder` implements the `_transform_pandas` method. So, when `MultiHotEncoder` transforms a dataset, it converts Arrow blocks to pandas DataFrames. During the conversion, lists are converted to ndarrays. This causes `MultiHotEncoder` to error, because `MultiHotEncoder` doesn't support ndarrays (only built-in lists).

## Related issue number

Fixes #31353 

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by Yard1 — APPROVED** [review]

File: python/ray/data/preprocessors/encoder.py [hunk 1]
```diff
@@ -324,7 +325,9 @@ def _transform_pandas(self, df: pd.DataFrame):
         _validate_df(df, *self.columns)
```
Comment (jiaodong) [inline python/ray/data/preprocessors/encoder.py:325]: if the problem is existing encoder cannot handle arrow data format that leads to numpy input format, shouldn't we implement `_transform_numpy` for this preprocessor instead ? Then in our existing preprocessor / batch mapper / predictor code, we will be able to locally pick the numpy path based on input format and preprocessor implementations.

Enhancing `_transform_pandas` here still introduces copies of input data between pandas <> arrow

Comment (bveeramani) [reply python/ray/data/preprocessors/encoder.py:325]: > if the problem is existing encoder cannot handle arrow data format that leads to numpy input format, shouldn't we implement _transform_numpy for this preprocessor instead ?

I'm not sure.

I wouldn't say that the Arrow data format leads to a NumPy input format. It's still tabular data; it's just that lists are encoded as NumPy arrays. Since the data represents tabular data, I think `_transform_pandas` might be easier to implement.

```
>>> df = pd.DataFrame({
...        "movie": ["spam", "ham"],
...        "genre": [["action"], ["comedy", "romance"]]
... })
>>> type(df["genre"][0])
<class 'list'>
>>> df
  movie              genre
0  spam           [action]
1   ham  [comedy, romance]
```

```
>>> df = pa.table(df).to_pandas()
>>> type(df["genre"][0])
<class 'numpy.ndarray'>
>>> df
  movie              genre
0  spam           [action]
1   ham  [comedy, romance]
```

> Enhancing _transform_pandas here still introduces copies of input data between pandas <> arrow

Which copy are you referring to? Converting an Arrow Table to a pandas DataFrame?

Comment (amogkam) [reply python/ray/data/preprocessors/encoder.py:325]: Agree with @bveeramani here- the problem is that the actual item of a particular row is a numpy array and not a list. Even with numpy path, this would still be a problem (the input would look like this: `{"column": np.array([np.array(["A", "B"])])}`). It's the inner numpy array that's troublesome.

We can still implement a `_transform_numpy` path, but as @bveeramani mentioned, given that this preprocessor is meant for tabular data and not tensors, not sure how much perf value we will get from this.

Regardless, this change is needed to make the pandas path work well.

**Review by amogkam — APPROVED** [review]

File: python/ray/data/preprocessors/encoder.py (hunk 1 above)
Comment (jiaodong) [reply python/ray/data/preprocessors/encoder.py:325]: i see, from description of the PR, where does `it converts Arrow blocks to pandas DataFrames` happen ? I was under the impression that you're suggesting original data format is arrow but we only have `_transform_pandas` in this component, in this case adding `_transform_numpy` will be preferred.

But if original data format is tabular, enhancing `_transform_pandas` is the right way to go.  

As a follow up I would like to understand the workload you're optimizing for e2e, since the snippet you had above
```python
df = pa.table(df).to_pandas()
```
does't have to call `to_pandas` if we don't need to, right ? What use cases we will start with an arrow dataset ?

Comment (bveeramani) [reply python/ray/data/preprocessors/encoder.py:325]: >where does it converts Arrow blocks to pandas DataFrames happen ?

I think `map_batches`

https://github.com/ray-project/ray/blob/892b4f0de2c2311fbb6f2a3989fc4e22cb8994b2/python/ray/data/dataset.py#L596-L598

eventually calls `pa.Table.to_pandas`: 

https://github.com/ray-project/ray/blob/892b4f0de2c2311fbb6f2a3989fc4e22cb8994b2/python/ray/data/_internal/arrow_block.py#L222-L229

>  I was under the impression that you're suggesting original data format is arrow but we only have _transform_pandas in this component, in this case adding _transform_numpy will be preferred.

I think this is the case? The block type is Arrow, and `MultiHotEncoder` only implements `_transform_pandas`. That said, the dataset represents tabular data.

> As a follow up I would like to understand the workload you're optimizing for e2e, since the snippet you had above

Let's say you have a dataset of Tweets, and each Tweet has a "Tags" column that represents a list of tags. In this case,  `MultiHotEncoder` doesn't work:
```
dataset = ray.data.read_parquet("s3://tweets") # Arrow dataset that represents tabular data
preprocessor = MultiHotEncoder(columns=["Tags"])
preprocessor.fit_transform(dataset) # This errors
```
